### DISCUSSION TITLE
prepare and check handle are not executed as expected on windows

### DISCUSSION THREAD

**Discussion by jack2007** [libuv/libuv#3550] [General]
Hi, I'm trying prepare and check handle as http://docs.libuv.org/en/v1.x/design.html.
I'm running a tcp server.
I want insert my logic in prepare callback before my read callback, and insert check callback after read callback.
On Linux, it works perfect as expected. 
But on windows, I get result that prepare callback is called after read callback, and check callback is called before read callback.

I go through the uv_run() windows code, 
https://github.com/libuv/libuv/blob/df63ee1a1de6d99821c172ea9590cba72917e718/src/win/core.c#L605

it looks like that on windows read callback is call in uv_process_reqs, uv_prepare_invoke() are behind uv_process_reqs. uv__poll on windows does not call read callback, but just insert i/o request into request queue.

Best regards.

**Comment by bnoordhuis** [comment]
I've converted this to a discussion because it sounds like a misunderstanding on your part (or maybe the documentation is unclear) but not a libuv bug.

> I want insert my logic in prepare callback before my read callback, and insert check callback after read callback.

The documentation on them says this:

- Prepare handles will run the given callback once per loop iteration, right before polling for i/o.
- Check handles will run the given callback once per loop iteration, right after polling for i/o.

I suspect you're conflating "polling" with "reading" but on Windows it's really "checking if something was read."

Libuv on Windows uses completion-based I/O (Linux is readiness-based) so that means you can't really make assumptions when reads or writes take place.

**Comment by vtjnash** [comment]
This looks likely to be the same underlying complaint as your PR https://github.com/libuv/libuv/pull/3585 was intending to "fix" in nodejs. So is nodejs also misunderstanding correct usage of libuv, or do we want to fix this in libuv? The upside to fixing this in libuv is that we remove a platform-specific discrepancy in the tests.

```diff
diff --git a/src/unix/core.c b/src/unix/core.c
index edd457a1..1aca087f 100644
--- a/src/unix/core.c
+++ b/src/unix/core.c
@@ -390,6 +390,11 @@ int uv_run(uv_loop_t* loop, uv_run_mode mode) {
       timeout = uv__backend_timeout(loop);
 
     uv__io_poll(loop, timeout);
+    /* Process immediate callbacks (e.g. write_cb) a small fixed number of
+     * times to avoid loop starvation.*/
+    for (r = 0; r < 16; r++)
+      if (!uv__process_reqs(loop))
+        break;
 
     /* Run one final update on the provider_idle_time in case uv__io_poll
      * returned because the timeout expired, but no events were received. This
diff --git a/src/win/core.c b/src/win/core.c
index c08dedc4..9ce0894f 100644
--- a/src/win/core.c
+++ b/src/win/core.c
@@ -614,6 +614,11 @@ int uv_run(uv_loop_t *loop, uv_run_mode mode) {
       uv__poll(loop, timeout);
     else
       uv__poll_wine(loop, timeout);
+    /* Process immediate callbacks (e.g. write_cb) a small fixed number of
+     * times to avoid loop starvation.*/
+    for (r = 0; r < 16; r++)
+      if (!uv__process_reqs(loop))
+        break;
 
     /* Run one final update on the provider_idle_time in case uv__poll*
      * returned because the timeout expired, but no events were received. This
diff --git a/test/test-pipe-set-non-blocking.c b/test/test-pipe-set-non-blocking.c
index 827e7264..c7804609 100644
--- a/test/test-pipe-set-non-blocking.c
+++ b/test/test-pipe-set-non-blocking.c
@@ -102,8 +102,7 @@ TEST_IMPL(pipe_set_non_blocking) {
     ASSERT(n == UV_EAGAIN); /* E_NOTIMPL */
     ASSERT(0 == uv_write(&write_req, (uv_stream_t*) &pipe_handle, &buf, 1, write_cb));
     ASSERT_NOT_NULL(write_req.handle);
-    ASSERT(1 == uv_run(uv_default_loop(), UV_RUN_ONCE)); /* queue write_cb */
-    ASSERT(0 == uv_run(uv_default_loop(), UV_RUN_ONCE)); /* process write_cb */
+    ASSERT(0 == uv_run(uv_default_loop(), UV_RUN_ONCE));
     ASSERT_NULL(write_req.handle); /* check for signaled completion of write_cb */
     n = buf.len;
 #endif
diff --git a/test/test-spawn.c b/test/test-spawn.c
index dfd5458e..a440423a 100644
--- a/test/test-spawn.c
+++ b/test/test-spawn.c
@@ -1675,9 +1675,6 @@ TEST_IMPL(closed_fd_events) {
   ASSERT(req.result == 1);
   uv_fs_req_cleanup(&req);
 
-#ifdef _WIN32
-  ASSERT(1 == uv_run(uv_default_loop(), UV_RUN_ONCE));
-#endif
   ASSERT(0 == uv_run(uv_default_loop(), UV_RUN_ONCE));
 
   /* should have received just one byte */
```

(n.b. we cannot loop until empty without causing a regression of https://github.com/libuv/libuv/pull/1787)

**Reply by vtjnash** [reply]
(fwiw, we only need exactly one call in src/win/core.c to fix this and align this expectation with the unix behavior. But I then put it in a loop so that other callbacks would also be flushed many times, so that all platforms would be more consistent in their use of the pending queue.)

**Reply by bnoordhuis** [reply]
I didn't understand @jack2007's as saying it's a regression from older libuv versions, rather that...

> insert my logic in prepare callback before my read callback, and insert check callback after read callback

...is something that works on Unices but not on Windows because of readiness vs. completion-based I/O. Node's old bump allocator exploited that quirk (on Unices) to reduce its memory footprint.

**Reply by vtjnash** [reply]
It is not a regression, since that is reasonable behavior. But nodejs is (incorrectly?) reliant upon the unix behavior of the order of events, as we saw recently in #3585
